File: collective-intelligence-frontend/src/services/decision.ts
// src/services/decision.ts
import axios from 'axios';

// Types
export interface DecisionProcess {
  id: string;
  discussion_id: string;
  title: string;
  status: string;
  process_template: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface DecisionStage {
  id: string;
  process_id: string;
  name: string;
  description: string;
  order_index: number;
  status: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface DecisionDocument {
  id: string;
  process_id: string;
  title: string;
  content: string;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface DecisionMetric {
  id: string;
  process_id: string;
  metric_name: string;
  metric_value: number;
  calculated_at: string;
}

// API functions
export const decisionApi = {
  // Get decision process for a discussion
  getDecisionProcess: (discussionId: string) => 
    axios.get<{ process: DecisionProcess, stages: DecisionStage[] }>(
      `/api/discussions/${discussionId}/decision-process`
    ),
  
  // Create a new decision process
  createDecisionProcess: (discussionId: string, title: string, template?: string) => 
    axios.post<{ process: DecisionProcess, stages: DecisionStage[] }>(
      `/api/discussions/${discussionId}/decision-process`,
      { title, template }
    ),
  
  // Update a decision stage
  updateDecisionStage: (stageId: string, status: string) => 
    axios.patch<DecisionStage>(`/api/decision-stages/${stageId}`, { status }),
  
  // Get decision document
  getDecisionDocument: (processId: string) => 
    axios.get<{ document: DecisionDocument }>(`/api/decision-processes/${processId}/document`),
  
  // Create decision document
  createDecisionDocument: (processId: string, title: string, content: string) => 
    axios.post<{ document: DecisionDocument }>(
      `/api/decision-processes/${processId}/document`,
      { title, content }
    ),
  
  // Update decision document
  updateDecisionDocument: (documentId: string, title: string, content: string) => 
    axios.put<{ document: DecisionDocument }>(
      `/api/decision-documents/${documentId}`,
      { title, content }
    ),
  
  // Get decision metrics
  getDecisionMetrics: (processId: string) => 
    axios.get<{ metrics: DecisionMetric[] }>(`/api/decision-processes/${processId}/metrics`)
};

File: backend/app/models/decision.py
# app/models/decision.py
import uuid
from datetime import datetime

from app import db

class DecisionProcess(db.Model):
    __tablename__ = 'decision_processes'
    
    id = db.Column(db.String(36), primary_key=True)
    discussion_id = db.Column(db.String(36), db.ForeignKey('discussions.id'), nullable=False)
    title = db.Column(db.String(200), nullable=False)
    status = db.Column(db.String(50), default='in_progress')
    process_template = db.Column(db.String(100), nullable=True)
    started_at = db.Column(db.DateTime, default=datetime.utcnow)
    completed_at = db.Column(db.DateTime, nullable=True)
    
    # Relationships
    stages = db.relationship('DecisionStage', back_populates='process', cascade='all, delete-orphan')
    documents = db.relationship('DecisionDocument', back_populates='process', cascade='all, delete-orphan')
    
    def __init__(self, discussion_id, title, process_template=None):
        self.id = str(uuid.uuid4())
        self.discussion_id = discussion_id
        self.title = title
        self.process_template = process_template
    
    def to_dict(self):
        return {
            'id': self.id,
            'discussion_id': self.discussion_id,
            'title': self.title,
            'status': self.status,
            'process_template': self.process_template,
            'started_at': self.started_at.isoformat(),
            'completed_at': self.completed_at.isoformat() if self.completed_at else None
        }

class DecisionStage(db.Model):
    __tablename__ = 'decision_stages'
    
    id = db.Column(db.String(36), primary_key=True)
    process_id = db.Column(db.String(36), db.ForeignKey('decision_processes.id'), nullable=False)
    name = db.Column(db.String(100), nullable=False)
    description = db.Column(db.Text)
    order_index = db.Column(db.Integer, nullable=False)
    status = db.Column(db.String(50), default='pending')
    started_at = db.Column(db.DateTime, nullable=True)
    completed_at = db.Column(db.DateTime, nullable=True)
    
    # Relationships
    process = db.relationship('DecisionProcess', back_populates='stages')
    
    def __init__(self, process_id, name, description, order_index):
        self.id = str(uuid.uuid4())
        self.process_id = process_id
        self.name = name
        self.description = description
        self.order_index = order_index
    
    def to_dict(self):
        return {
            'id': self.id,
            'process_id': self.process_id,
            'name': self.name,
            'description': self.description,
            'order_index': self.order_index,
            'status': self.status,
            'started_at': self.started_at.isoformat() if self.started_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None
        }

class DecisionDocument(db.Model):
    __tablename__ = 'decision_documents'
    
    id = db.Column(db.String(36), primary_key=True)
    process_id = db.Column(db.String(36), db.ForeignKey('decision_processes.id'), nullable=False)
    title = db.Column(db.String(200), nullable=False)
    content = db.Column(db.Text, nullable=False)
    version = db.Column(db.Integer, default=1)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    process = db.relationship('DecisionProcess', back_populates='documents')
    
    def __init__(self, process_id, title, content, version=1):
        self.id = str(uuid.uuid4())
        self.process_id = process_id
        self.title = title
        self.content = content
        self.version = version
    
    def to_dict(self):
        return {
            'id': self.id,
            'process_id': self.process_id,
            'title': self.title,
            'content': self.content,
            'version': self.version,
            'created_at': self.created_at.isoformat(),
            'updated_at': self.updated_at.isoformat()
        }